pkgs/tools/typesetting/tex/texlive/default.nix

```
@@ -34,392 +34,8 @@ let
-
-      # add them elsewhere so that collections cover all packages
-      collection-metapost.deps = orig.collection-metapost.deps ++ [ "metafont" ];
-      collection-plaingeneric.deps = orig.collection-plaingeneric.deps ++ [ "xdvi" ];
-
-      #### misc
-
-      # tlpdb lists license as "unknown", but the README says lppl13: http://mirrors.ctan.org/language/arabic/arabi-add/README
-      arabi-add.license = [  "lppl13c" ];
-
-      # TODO: remove this when updating to texlive-2023, npp-for-context is no longer in texlive
-      # tlpdb lists license as "noinfo", but it's gpl3: https://github.com/luigiScarso/context-npp
-      npp-for-context.license = [  "gpl3Only" ];
-
-      texdoc = {
-        extraRevision = "-tlpdb${toString tlpdbVersion.revision}";
-        extraVersion = "-tlpdb-${toString tlpdbVersion.revision}";
-
-        # build Data.tlpdb.lua (part of the 'tlType == "run"' package)
-        postUnpack = ''
-          if [[ -f "$out"/scripts/texdoc/texdoc.tlu ]]; then
-            unxz --stdout "${tlpdbxz}" > texlive.tlpdb
-
-            # create dummy doc file to ensure that texdoc does not return an error
-            mkdir -p support/texdoc
-            touch support/texdoc/NEWS
-
-            TEXMFCNF="${bin.core}"/share/texmf-dist/web2c TEXMF="$out" TEXDOCS=. TEXMFVAR=. \
-              "${bin.luatex}"/bin/texlua "$out"/scripts/texdoc/texdoc.tlu \
-              -c texlive_tlpdb=texlive.tlpdb -lM texdoc
-
-            cp texdoc/cache-tlpdb.lua "$out"/scripts/texdoc/Data.tlpdb.lua
-          fi
-        '';
-      };
-
-      "texlive.infra" = {
-        extraRevision = ".tlpdb${toString tlpdbVersion.revision}";
-        extraVersion = "-tlpdb-${toString tlpdbVersion.revision}";
-
-        # add license of tlmgr and TeXLive::* perl packages and of bin.core
-        license = [ "gpl2Plus" ] ++ lib.toList bin.core.meta.license.shortName ++ orig."texlive.infra".license or [ ];
-
-        scriptsFolder = "texlive";
-        extraBuildInputs = [ coreutils gnused gnupg (lib.last tl.kpathsea.pkgs) (perl.withPackages (ps: with ps; [ Tk ])) ];
-
-        # make tlmgr believe it can use kpsewhich to evaluate TEXMFROOT
-        postFixup = ''
-          substituteInPlace "$out"/bin/tlmgr \
-            --replace 'if (-r "$bindir/$kpsewhichname")' 'if (1)'
-          sed -i '2i$ENV{PATH}='"'"'${lib.makeBinPath [ gnupg ]}'"'"' . ($ENV{PATH} ? ":$ENV{PATH}" : '"'''"');' "$out"/bin/tlmgr
-          sed -i '2iPATH="${lib.makeBinPath [ coreutils gnused (lib.last tl.kpathsea.pkgs) ]}''${PATH:+:$PATH}"' "$out"/bin/mktexlsr
-        '';
-
-        # add minimal texlive.tlpdb
-        postUnpack = ''
-          if [[ "$tlType" == "tlpkg" ]] ; then
-            xzcat "${tlpdbxz}" | sed -n -e '/^name \(00texlive.config\|00texlive.installation\)$/,/^$/p' > "$out"/texlive.tlpdb
-          fi
-        '';
-      };
-    }; # overrides
+    overrides = callPackage ./tlpdb-overrides.nix { inherit bin tlpdb tlpdbxz tl; };
```

why not just
```suggestion
    overrides = import ./tlpdb-overrides.nix { ... } { inherit bin tlpdb tlpdbxz tl; };
```
in this case? given we don't have a scope yet at this stage

---

Another concern is that each `callPackage` introduces an additional `override`. I am thinking for instance that `perl` should probably be the same across all of `texlive`, and so only overrideable from `texlive.override`.

---

I see your point, This isn't really a concern for all the other scopes where overriding things outside of the scope for the whole scope isn't really a thing that makes sense, but it might make sense for texlive.

It should still be possible to override the inputs with something like
```
texlive.overrideScope (self: super: { callPackage = super.newScope { perl = ... ; }; })
```
or by just inserting perl into the scope with
```
texlive.overrideScope (self: super: { perl = ...; })
```
But this isn't a pattern that I have seen anywhere yet and would probably require some explicit documentation.

---

> ```
> texlive.overrideScope (self: super: { perl = ...; })
> ```

Wouldn't it be essentially equivalent to `texlive.override { perl = ...; }`? I was picturing `texlive.override` for external overrides, `texlive.overrideScope` for internal overrides. Plus, overriding `perl` in `overrideScope` would apply to `self.callPackage`, not `callPackage`, I think?

---

I think we might have talked past each other.

If the texlive toplevel explicitly lists perl as one of its arguments and manually distributes this argument to the other functions/files (like tlpdb-overrides.nix) as it's currently done, it is possible to change the perl used throughout texlive with `texlive.override { perl = ... }`.

However, in #250626 this no longer is case. There, the perl argument expected by tlpdb-overrides.nix is provided by `texlive.callPackage`. `texlive.callPackage` provides the arguments tlpdb-overrides.nix expects from an attrset that should (I don't know how to verify) look something like
```
pkgs // pkgs.xorg 
     // pkgs.texlive 
     // { inherit (args-to-tl) tlpdb tlpdbxzHash src version mirrors useFixedHashes fixedHashes }
     // { ghostscript = pkgs.ghostscript_headless; harfbuzz = pkgs.harfbuzz.override { withIcu = true; withGraphite2 = true; };
```
The `pkgs // pkgs.xorg` comes from `pkgs.newScope`,` pkgs.texlive` comes from the `makeScopeWithSplicing` in texlive and the remaining args are added manually in 
https://github.com/NixOS/nixpkgs/blob/4833e3883299f02c463c1afd2d5667481e8117e3/pkgs/tools/typesetting/tex/texlive/default.nix#L36-L39

This implies that the perl passed by `texlive.callPackage` to `tlpdb-overrides.nix` can by overridden by putting a perl attribute inside `pkgs.texlive` with `texlive.overrideScope (self: super: { perl = ...; })`:

```
Added 20063 variables.
nix-repl> getPerl = { perl }: perl                                                 

nix-repl> texlive.callPackage getPerl {}                                           
«derivation /nix/store/kpfm50yy4r21h5vm5666ll84ad3g7rj4-perl-5.38.0.drv»

nix-repl> (texlive.overrideScope (_: _: { perl = "foo"; })).callPackage getPerl {}
"foo"
```

---

> I am overthinking this, aren't I?

Overthinking this is definitely better than ending up with a terrible interface we cannot easily change. Texlive doesn't really feel like it should be a scope, so having to use `.overrideScope` to override one of the "external" dependencies feels really weird and I'm reasonably confident that no one would try using `.overrideScope` without digging into the source. Overriding things in nixpkgs already is terribly inconsistent, we really should try not to make it worse.

---

If we really want texlive to be more like the other language ecoysystems, `pkgs.texlive` should be a single derivation  containing everything built from the texlive source tarball and the package set would be implemented as an passthru attribute of it.
But doing this while maintaining compability is probably impossible (or at the very least a lot of work) and then we'd end up with a `pkgs.texlive` derivation that isn't really usable on its own, what also isn't what users would expect.

---

To be clear, I was mostly suggesting collapsing `texlive.bin` into a single derivation with many outputs that would take the place of the interpreter as it allows directly copying the structure of the other language environments without having to deal with the weird situation where we try to copy their interface while having a completely different (and far more complicated) internal structure -- texlive is a remarkably tiny scope where the two members bin and pkgs awkwardly depend on each other. I haven't thought about this (or even tried to implement it) too much, but at first glance this appears like a major simplification without any major drawbacks other than the "interpreter derivation" `pkgs.texlive` having awkward contents for its name. Anyway, this could also be explored some time in the future when the interface with the new top levels is in place and it really only is an implementation detail.

I really like your idea for the external interface by the way.

---

> I was mostly suggesting collapsing `texlive.bin` into a single derivation with many outputs that would take the place of the interpreter

I thought of that too, but somehow it does not fit here as you can't use `texlive` without an environment. I do wonder about the rationale behind the split `core`/`core-big`. Was it about closure size? Compilation time? Incidentally, the multi-output derivations have the same problem now: if you override a script in the `tex` output, it won't take because `out` does not get overridden (until we stabilize `buildTeXLivePackage`, which is the correct spot for doing an override).

Since easy overriding will be a new thing, would it be enough to put a huge disclaimer that the internal `bin.core` will be reviewed and may change before the next release? Take some pressure off of getting every little detail right in one go.

---

> I thought of that too, but somehow it does not fit here as you can't use texlive without an environmen

Fully agree. The best we could do to mitigate the problem is naming it `texlive-bin` (after all, `texlive-bin.withPackages (p: [ ... ])` does make sense) while adding the new `texlivePackages`, `texliveFull` etc toplevels and keeping `texlive` for a compat layer. But this would still feel a little weird imho.

> I do wonder about the rationale behind the split core/core-big.

I checked the log and it has been like that ever since the rewrite in https://github.com/NixOS/nixpkgs/commit/22fd6e1e4c52f8ab0651409812f9c53a59c1077c. I can't find a PR or a discussion on the mailing list about that.

> Was it about closure size? Compilation time? 

I don't see either of these as a issue as different outputs yield different store-paths that can also be substituted individually. And if anything, the cumulated build time of the separate derivations should be higher than if there was only one as the configure scripts has to be run multiple times. There only is a benefit if someone doesn't use the cache and wants a minimal installation or when cross-compiling (But even compiling `bin.core` just for tangle is rather wasteful, so it probably wasn't too much of a concern) -- And both of these are imho unimportant edge cases.

> Since easy overriding will be a new thing, would it be enough to put a huge disclaimer that the internal bin.core will be reviewed and may change before the next release? Take some pressure off of getting every little detail right in one go.

Setting the expectations right is always a good idea. We could also discourage external uses of `texlive.bin` by adding an eval-warning to the attribute in the compat fixups once external users can use `texlivePackages.foo`. (This still requires some changes though. Some pdf viewers depend on `texlive.bin.core` for libsynctex, which is not contained in the `synctex` bin container.)

---

> The best we could do to mitigate the problem is naming it `texlive-bin` (after all, `texlive-bin.withPackages (p: [ ... ])` does make sense) while adding the new `texlivePackages`, `texliveFull` etc toplevels and keeping `texlive` for a compat layer. But this would still feel a little weird imho.

I thought about this more carefully, and we really can't split `texlive-bin` off of the package set. The package set contains all the non-binary stuff, like scripts, documentation, man pages of the core packages. So we would have to generate those containers ourselves, but for little benefit – who would want to upgrade the entire package set to the next year, but keep the old binaries?

Between this, and some experiments I have done today, I am convincing myself that we want top level `texlive`, `texliveYYYY` (essentially aliases of the infraonly scheme, the first one exposing `.combine`), `texliveSchemeAAAA`, `texliveYYYYSchemeAAA` (possibly without 'Scheme'?). Each top level attribute exports the packages directly underneath, with `.pkgs` visible in case of name clashes. At which point, we could export the aliases `texlivePackages`, `texliveYYYYPackages` but for the sole purpose of _searching_.

Actual usage would look like:
```nix
buildInputs = [ texliveSchemeSmall ] ++ (with texliveSchemeSmall; [ libertinus-fonts ]);
texliveSchemeSmall.withPackages (ps: with ps; [ libertinus-fonts ]) # for fontconfig
texliveSchemeSmall.override { extraName = ...; paperSize = ...; }
texliveSchemeSmall.overrideScope (final: prev: ...)
```

Re binaries, let's not forget to open an issue about reviewing `bin.core` after we are done with the rest, and check how we export the libraries like kpathsea, synctex (maybe a `lib` container?).